pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,647 @@
+
+	if !okOld || !okNew {
+		// typecasting of one failed, handle this better, may be log entry
+		return
+	}
+	// if the change in schedule results in next requeue having to be sooner than it already was,
+	// it will be handled here by the queue. If the next requeue is further than previous schedule,
+	// the sync loop will essentially be a no-op for the already queued key with old schedule.
+	if oldCJ.Spec.Schedule != newCJ.Spec.Schedule {
+		// schedule changed, change the requeue time
+		sched, err := cron.ParseStandard(newCJ.Spec.Schedule)
+		if err != nil {
+			// this is likely a user error in defining the spec value
+			// we should log the error and not reconcile this cronjob until an update to spec
+			klog.V(2).InfoS("unparseable schedule for cronjob", "cronjob", klog.KRef(newCJ.GetNamespace(), newCJ.GetName()), "schedule", newCJ.Spec.Schedule, "err", err)
+			jm.recorder.Eventf(newCJ, corev1.EventTypeWarning, "UnParseableCronJobSchedule", "unparseable schedule for cronjob: %s", newCJ.Spec.Schedule)
+			return
+		}
+		now := jm.now()
+		t := nextScheduledTimeDuration(sched, now)
+
+		jm.enqueueControllerWithTime(curr, *t)
+		return
+	}
+
+	// other parameters changed, requeue this now and if this gets triggered
+	// within deadline, sync loop will work on the CJ otherwise updates will be handled
+	// during the next schedule
+	// TODO: need to handle the change of spec.JobTemplate.metadata.labels explicitly
+	//   to cleanup jobs with old labels
+	jm.enqueueController(curr)
+}
+
+// TODO: @alpatel we need to return errors from this function, in order to allow
+//  for errors to propagate up the reconcile function and kick in queues exponential
+//  backed off retries more details on inline code comments in the func
+
+// syncOne reconciles a CronJob with a list of any Jobs that it created.
+// All known jobs created by "cj" should be included in "js".
+// The current time is passed in to facilitate testing.
+// It has no receiver, to facilitate testing.
+func syncOne2(
+	cj *batchv1beta1.CronJob,
+	js []batchv1.Job,
+	now time.Time,
+	jc jobControlInterface,
+	cjc cjControlInterface,
+	recorder record.EventRecorder) (*time.Duration, error) {
+
+	childrenJobs := make(map[types.UID]bool)
+	for _, j := range js {
+		childrenJobs[j.ObjectMeta.UID] = true
+		found := inActiveList(*cj, j.ObjectMeta.UID)
+		if !found && !IsJobFinished(&j) {
+			recorder.Eventf(cj, corev1.EventTypeWarning, "UnexpectedJob", "Saw a job that the controller did not create or forgot: %s", j.Name)
+			// We found an unfinished job that has us as the parent, but it is not in our Active list.
+			// This could happen if we crashed right after creating the Job and before updating the status,
+			// or if our jobs list is newer than our cj status after a relist, or if someone intentionally created
+			// a job that they wanted us to adopt.
+
+			// TODO: maybe handle the adoption case?  Concurrency/suspend rules will not apply in that case, obviously, since we can't
```

Comment:
I doubt this use-case is common enough to warrant the concern.

Reply:
+1 removed the TODO